Sources/SourceKitLSP/Swift/SwiftLanguageServer.swift

```diff
@@ -161,6 +161,71 @@ public final class SwiftLanguageServer: ToolchainLanguageServer {
     }
   }
 
+  /// Update lexical and syntactic tokens for the given `snapshot`.
+  func updateLexicalAndSyntacticTokens(
+    response: SKDResponseDictionary,
+    for snapshot: DocumentSnapshot
+  ) {
+    let uri = snapshot.document.uri
+
+    if let syntaxMap: SKDResponseArray = response[keys.syntaxmap] {
+      let tokenParser = SemanticTokenParser(
+        sourcekitd: sourcekitd,
+        snapshot: snapshot
+      )
+      let tokens = tokenParser.parseTokens(syntaxMap)
+      do {
+        try documentManager.addLexicalTokens(uri, tokens: tokens)
+      } catch {
+        log("updating lexical tokens for \(uri) failed: \(error)", level: .warning)
+      }
+    }
+
+    if let substructure: SKDResponseArray = response[keys.substructure] {
+      let tokenParser = SemanticTokenParser(
+        sourcekitd: sourcekitd,
+        snapshot: snapshot,
+        useName: true
+      )
+      let tokens = tokenParser.parseTokens(substructure)
+      do {
+        try documentManager.replaceSyntacticTokens(uri, tokens: tokens)
+      } catch {
+        log("updating lexical tokens for \(uri) failed: \(error)", level: .warning)
+      }
+    }
+  }
+
+  /// Update semantic tokens for the given `snapshot`.
+  func updateSemanticTokens(
+    response: SKDResponseDictionary,
+    for snapshot: DocumentSnapshot
+  ) {
+    let uri = snapshot.document.uri
+    guard let skTokens: SKDResponseArray = response[keys.annotations] else {
+      return
+    }
+
+    let tokenParser = SemanticTokenParser(
+      sourcekitd: sourcekitd,
+      snapshot: snapshot
+    )
+    let tokens = tokenParser.parseTokens(skTokens)
+
+    do {
+      try documentManager.replaceSemanticTokens(uri, tokens: tokens)
+      if clientCapabilities.workspace?.semanticTokens?.refreshSupport ?? false {
```

Comment:
I would move the code to inform the client to a new function. Right now, we assume that `updateSemanticTokens` is called after `updateLexicalAndSyntacticTokens` to transfer the syntax information to the client. But that doesn’t necessarily need to be the case, right?

Reply:
I've moved the refresh logic into a separate logic. Clients (notably VSCode) however don't _need_ these refresh requests, they automatically request tokens after every edit, so we get responsive highlighting one way or the other. See also the discussion below regarding testability.

Since semantic tokens are updated asynchronously, I think it makes sense for them to cause a refresh request, this doesn't apply to the syntactic/lexical tokens however.

Reply:
I don’t think it’s an issue right now because `SwiftLanguageServer` serializes all requests on its queue, but I’m a little worried that if we should drop that guarantee, we’ll end up in a race condition where we process the semantics tokens request before getting the response from the edit.

Also, after looking at the documentation of `WorkspaceSemanticTokensRefreshRequest` some more, sending it every time we re-computed semantic tokens seems a bit of an over-kill to me, as it causes a *workspace-wide* refresh of all semantic tokens. IIUC the LSP idea for this request would be that we don’t send the response until we actually have the *semantic* tokens for that file. I am really not sure how to approach this best and would like to hear your opinion. As far as I see it, we have the following options:

* Leave the implementation as-is, providing syntactic highlighting as LSP’s semantic highlight directly and sending a `WorkspaceSemanticTokensRefreshRequest` once semantic tokens are available
  * Advantages:
    * We directly provide syntactic highlighting information to the editor
    * Changes in one file reflect as semantic highlighting changes in the other files. E.g. if file A declares `struct Foo { static func bar() } }` and file B calls `Foo.bar()`, we rename `bar -> baz` in file A, then file B will stop coloring `bar` as a member. We don’t seem to have this support in Xcode at the moment either, so I would say that it’s not a super important feature to have
  * Disadvantages:
    *  Potentially impacts performance because we request a *project-wide* refresh of semantic tokens, which stops the editor from doing more local, scoped requests (which starts to become relevant as soon as we implement ranged semantic highlighting)
* Only reply to the semantic tokens request once we actually have semantic tokens
  * Advantages:
    * I think this is closer to how LSP envisions semantic tokens support
    * Less of a performance-impact
  * Disadvantages:
    * No way to provide our syntactic highlighting information until we have semantic tokens
    * We would need to have at least some limited concurrent request support in SourceKit-LSP, I’m not sure how hard that would be to implement.
  * Questions:
    * Would we still want to merge the information with our syntactic highlighting information or leave the syntactic highlighting to the editor?

Reply:
We'll keep it like this (using refresh requests) for now, since it seems to work correctly and leave callback-based handling up for future consideration, as discussed.